core/src/main/java/io/grpc/internal/CallCredentialsApplyingTransportFactory.java

```diff
@@ -126,5 +140,50 @@ public Attributes getTransportAttrs() {
         return delegate.newStream(method, headers, callOptions);
       }
     }
+
+    @Override
+    public Runnable start(Listener listener) {
+      this.listener = new CallCredentialsApplyingTransportListener(listener);
+      return super.start(this.listener);
+    }
+    // TODO(zivy@): cancel pending appliers when shutdownNow.
+
+    private class CallCredentialsApplyingTransportListener implements Listener {
+      private final Listener delegateListener;
+      private volatile boolean savedTransportTerminated;
+
+      public CallCredentialsApplyingTransportListener(Listener listener) {
+        this.delegateListener = listener;
+      }
+
+      @Override
+      public void transportShutdown(Status s) {
+        delegateListener.transportShutdown(s);
+      }
+
+      @Override
+      public void transportTerminated() {
+        savedTransportTerminated = true;
+        maybeTerminated();
+      }
+
+      @Override
+      public void transportReady() {
+        delegateListener.transportReady();
+      }
+
+      @Override
+      public void transportInUse(boolean inUse) {
+        delegateListener.transportInUse(inUse);
+      }
+
+      public void maybeTerminated() {
+        synchronized (this) {
+          if (pendingApplier.get() == 0 && savedTransportTerminated) {
+            delegateListener.transportTerminated();
```

Comment:
I'd feel better if this was run without the lock held. Right now it looks very easy: just set a local variable while the lock is held if the callback should be run